driver-core/src/main/com/mongodb/client/model/expressions/MqlExpression.java

```diff
@@ -0,0 +1,122 @@
+import org.bson.codecs.configuration.CodecRegistry;
+
+import java.util.function.Function;
+
+final class MqlExpression<T extends Expression>
+        implements Expression, BooleanExpression, IntegerExpression, NumberExpression,
+        StringExpression, DateExpression, DocumentExpression, ArrayExpression<T> {
+
+    private final Function<CodecRegistry, BsonValue> fn;
+
+    protected MqlExpression(final Function<CodecRegistry, BsonValue> fn) {
+        this.fn = fn;
+    }
+
+    /**
+     * Exposes the evaluated BsonValue so that expressions may be used in
+     * aggregations. Non-public, as it is intended to be used only by the
+     * {@link MqlExpressionCodec}.
+     */
+    BsonValue toBsonValue(final CodecRegistry codecRegistry) {
+        return fn.apply(codecRegistry);
+    }
+
+    private Function<CodecRegistry, BsonValue> astDoc(final String name, final BsonDocument value) {
+        return (cr) -> new BsonDocument(name, value);
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name) {
+        return (cr) -> new BsonDocument(name, this.toBsonValue(cr));
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name, final Expression param1) {
+        return (cr) -> {
+            BsonArray value = new BsonArray();
+            value.add(this.toBsonValue(cr));
+            value.add(extractBsonValue(cr, param1));
+            return new BsonDocument(name, value);
+        };
+    }
+
+    private Function<CodecRegistry, BsonValue> ast(final String name, final Expression param1, final Expression param2) {
+        return (cr) -> {
+            BsonArray value = new BsonArray();
+            value.add(this.toBsonValue(cr));
+            value.add(extractBsonValue(cr, param1));
+            value.add(extractBsonValue(cr, param2));
+            return new BsonDocument(name, value);
+        };
+    }
+
+    /**
+     * Takes an expression and converts it to a BsonValue. MqlExpression will be
+     * the only implementation of Expression and all subclasses, so this will
+     * not mis-cast an expression as anything else.
+     */
+    @SuppressWarnings("rawtypes")
+    protected BsonValue extractBsonValue(final CodecRegistry cr, final Expression expression) {
+        return ((MqlExpression) expression).toBsonValue(cr);
+    }
+
+    /**
+     * Converts an MqlExpression to any subtype of Expression. Users must not
+     * extend Expression or its subtypes, so MqlExpression will implement any R.
+     */
+    @SuppressWarnings("unchecked")
+    <R extends Expression> R assertImplementsAllExpressions() {
+        return (R) this;
+    }
+
+    protected <R extends Expression> R newMqlExpression(final Function<CodecRegistry, BsonValue> ast) {
+        return new MqlExpression<>(ast).assertImplementsAllExpressions();
+    }
+
+    private <R extends Expression> R variable(final String variable) {
+        return newMqlExpression((cr) -> new BsonString(variable));
+    }
```

Comment:
I will leave both, they will be used in the next PR.

Reply:
I have two concerns:

- How reviewers are supposed to differentiate actually dead code from dead code that is expected to become live in conditionally on the future PR without wasting their time?
- What happens with this code if as a result of the preview process the code in the new PR gets changed, and this method does not become live? This code will not be seen as new in the next PR, so if the next PR stops using the method, it will be more likely that the dead method stays.

Reply:
Discussed in Zoom:

- If dead code is placed in a PR intentionally, knowing that the next PR will use it (for example, if there will be two next PRs crated concurrently, and both need the code), then the author leaves a comment about it to reviewers. If the code stays dead as a result of review changes in the next PRs, then the situation is indistinguishable from a situation when a PR changes existing code such that it makes some other existing code dead, and should be addressed in the same way.